compiler/src/dotty/tools/dotc/core/Annotations.scala

```diff
@@ -54,18 +55,26 @@ object Annotations {
      *  type, since ranges cannot be types of trees.
      */
     def mapWith(tm: TypeMap)(using Context) =
-      val args = arguments
+      val args = tpd.allArguments(tree)
       if args.isEmpty then this
       else
+        // Checks if `tm` would result in any change by applying it to types
+        // inside the annotations' arguments and checking if the resulting types
+        // are different.
         val findDiff = new TreeAccumulator[Type]:
           def apply(x: Type, tree: Tree)(using Context): Type =
             if tm.isRange(x) then x
             else
               val tp1 = tm(tree.tpe)
-              foldOver(if tp1 frozen_=:= tree.tpe then x else tp1, tree)
+              foldOver(if !tp1.exists || (tp1 frozen_=:= tree.tpe) then x else tp1, tree)
         val diff = findDiff(NoType, args)
         if tm.isRange(diff) then EmptyAnnotation
-        else if diff.exists then derivedAnnotation(tm.mapOver(tree))
+        else if diff.exists then
+          // If the annotation has been transformed, we need to make sure that the
```

Comment:
Why does transforming an annotation duplicate symbols? Don't we forget the old copy and use only the transformed one?

Reply:
The problem is that we have an `Annotated` tree whose type is an `AnnotatedType`, the `Annotated` tree has an `annot` field which is a tree and the `AnnotatedType` has `.annot.tree`. Initially, these tree are equal, but in PostTyper we call `transformAnnot` on both separately, se if PostTyper does some transformation, we end up with two non-referentially-equal trees that define the same symbol and crash in pickler (if PostTyper is the identity, then there is only shared tree and pickler is happy).

I think we could fix this locally in pickler by not pickling the `annot` field of the Annotated tree and instead reconstructing it from the type on unpickling, but more generally for type trees, it seems that `Annotation#annot` should just be a forward call to `.tpe.tree.annot`

Reply:
Actually I spoke too soon, it seems that in practice the `Annotated` trees are kept in sync with their  `AnnotatedType`. Things start going wrong when that `AnnotatedType` is used to type another tree (for example in https://github.com/scala/scala3/blob/99f431ef102b6991378003aae793e129f18e235e/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L153 but likely also just through type inference). Since there's no easy way to know if the same `AnnotatedType` is shared by multiple trees, this PR assumes this is always the case and defensively performs symbol copies.